maven-plugin/src/main/java/org/wildfly/galleon/maven/UserFeaturePackBuildMojo.java

```diff
@@ -0,0 +1,253 @@
+
+        for (String defaultPackage : buildConfig.getDefaultPackages()) {
+            fpBuilder.getSpecBuilder().addDefaultPackage(defaultPackage);
+        }
+
+        try {
+            processFeaturePackDependencies(buildConfig, fpBuilder.getSpecBuilder());
+        } catch (MojoExecutionException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new MojoExecutionException("Failed to process dependencies", e);
+        }
+
+        final Path srcModulesDir = resources.resolve(WfConstants.MODULES);
+        if (Files.exists(srcModulesDir)) {
+            addModulePackages(srcModulesDir, fpBuilder, resources);
+        }
+
+        buildFeaturePack(fpBuilder, buildConfig);
+    }
+
+    private void addModulePackages(final Path srcModulesDir, final FeaturePackDescription.Builder fpBuilder, final Path targetResources) throws MojoExecutionException {
+        final Map<String, Path> moduleXmlByPkgName = new HashMap<>();
+        try {
+            Util.findModules(srcModulesDir, moduleXmlByPkgName);
+            packageModules(fpBuilder, targetResources, moduleXmlByPkgName, null);
+        } catch (IOException e) {
+            throw new MojoExecutionException("Failed to process modules content", e);
+        }
+    }
+
+    private WildFlyFeaturePackBuild getBuildConfig() throws MojoExecutionException {
+        if (buildConfig == null) {
+            final Path path = Paths.get(configDir.getAbsolutePath(), configFile);
+            if (Files.exists(path)) {
+                buildConfig = Util.loadFeaturePackBuildConfig(path);
+            } else {
+                buildConfig = generateConfig();
+            }
+        }
+        return buildConfig;
+    }
+
+    private WildFlyFeaturePackBuild generateConfig() throws MojoExecutionException {
+        generate = true;
+        if (fpLocation == null) {
+            throw new MojoExecutionException("No feature-pack-location set");
+        }
+        WildFlyFeaturePackBuild.Builder builder = WildFlyFeaturePackBuild.builder();
+        builder.setProducer(FeaturePackLocation.fromString(fpLocation));
+        Map<FeaturePackLocation, Artifact> directs = retrieveDirectDependencies();
+        if (directs.isEmpty()) {
+            throw new MojoExecutionException("No dependency on WildFly feature-pack retrieved. "
+                    + "WildFly feature-pack must be a dependency of the project.");
+        }
+        Map<FeaturePackLocation, org.apache.maven.artifact.Artifact> transitives = retrieveTransitiveDependencies(directs.keySet());
+        for (Entry<FeaturePackLocation, Artifact> entry : directs.entrySet()) {
+            FeaturePackConfig.Builder depBuilder = FeaturePackConfig.builder(entry.getKey());
+            depBuilder.setInheritConfigs(false);
+            depBuilder.setInheritPackages(false);
+            Artifact a = entry.getValue();
+            debug("Adding %s dependency", a.getGroupId() + ":" + a.getArtifactId());
```

Comment:
groupId and artifactId should be a part of the format. The point of the format is to avoid unnecessary string concatenations.